src/wayland/virtual_keyboard/virtual_keyboard_handle.rs

```diff
@@ -169,54 +165,38 @@ where
 }
 
 /// Handle the zwp_virtual_keyboard_v1::keymap request.
-fn update_keymap<D>(data: &VirtualKeyboardUserData<D>, format: u32, fd: OwnedFd, size: usize)
+///
+/// The `true` returns when keymap was properly loaded.
+fn update_keymap<D>(data: &VirtualKeyboardUserData<D>, format: u32, fd: OwnedFd, size: usize) -> bool
 where
     D: SeatHandler + 'static,
 {
     // Only libxkbcommon compatible keymaps are supported.
     if format != KeymapFormat::XkbV1 as u32 {
         debug!("Unsupported keymap format: {format:?}");
-        return;
-    }
-
-    // Ignore potentially malicious requests.
-    if size > MAX_KEYMAP_SIZE {
-        debug!("Excessive keymap size: {size:?}");
-        return;
+        return false;
     }
 
-    // Read entire keymap.
-    let mut keymap_buffer = vec![0; size];
-    let mut file = File::from(fd);
-    if let Err(err) = file.read_exact(&mut keymap_buffer) {
-        debug!("Could not read keymap: {err}");
-        return;
-    }
-    let mut new_keymap = match String::from_utf8(keymap_buffer) {
-        Ok(keymap) => keymap,
-        Err(err) => {
-            debug!("Invalid utf8 keymap: {err}");
-            return;
-        }
-    };
-
-    // Ignore everything after the first nul byte.
-    if let Some(nul_index) = new_keymap.find('\0') {
-        new_keymap.truncate(nul_index);
-    }
-
-    // Attempt to parse the new keymap.
-    let new_keymap = xkb::Keymap::new_from_string(
-        &xkb::Context::new(xkb::CONTEXT_NO_FLAGS),
-        new_keymap,
-        xkb::KEYMAP_FORMAT_TEXT_V1,
-        xkb::KEYMAP_COMPILE_NO_FLAGS,
-    );
-    let new_keymap = match new_keymap {
-        Some(keymap) => keymap,
-        None => {
+    let context = xkb::Context::new(xkb::CONTEXT_NO_FLAGS);
+    let new_keymap = match unsafe {
+        xkb::Keymap::new_from_fd(
+            &context,
+            // SAFETY: We still need to close the FD afterwards, so using `as_raw` to map
+            // the fd.
+            fd.as_raw_fd(),
```

Comment:
xkbcommon 0.7 (which we just updated to) does take an `OwnedFd` here. Please rebase and remove the unsafe.

Reply:
I can't remove the `unsafe` though. The function is still `unsafe`.